src/mmrelay/migrate.py

```
@@ -0,0 +1,1883 @@
+        """
+        Create a MigrationError representing a failure of a specific migration step.
+
+        Parameters:
+            step (str): Name of the migration step that failed.
+            detail (str): Human-readable detail describing the failure.
+
+        Returns:
+            MigrationError: Error instance with message "<step> migration failed: <detail>".
+        """
+        exc = cls(f"{step} migration failed: {detail}")
+        setattr(exc, "step", step)
```

Ruff (0.14.14)

[warning] Do not call `setattr` with a constant attribute value. It is not any safer than normal property access.

Replace `setattr` with assignment

(B010)